asserts/snapasserts/validation_sets.go

```diff
@@ -449,3 +449,68 @@ func (v *ValidationSets) CheckInstalledSnaps(snaps []*InstalledSnap) error {
 	}
 	return nil
 }
+
+func (v *ValidationSets) constraintsForSnap(snapRef naming.SnapRef) *snapContraints {
+	if snapRef.ID() != "" {
+		return v.snaps[snapRef.ID()]
+	}
+	// snapID not available, find by snap name
+	for _, cstrs := range v.snaps {
+		if cstrs.name == snapRef.SnapName() {
+			return cstrs
+		}
+	}
+	return nil
+}
+
+// CheckPresenceRequired returns the list of all validation sets that declare
+// presence of the given snap as required and the required revision (or
+// snap.R(0) if no specific revision is required).
+// The method assumes that validation sets are not in conflict.
+func (v *ValidationSets) CheckPresenceRequired(snapRef naming.SnapRef) ([]string, snap.Revision) {
+	cstrs := v.constraintsForSnap(snapRef)
+	if cstrs == nil || cstrs.presence != asserts.PresenceRequired {
+		return nil, unspecifiedRevision
+	}
+
+	snapRev := unspecifiedRevision
+	var keys []string
+	for rev, revCstr := range cstrs.revisions {
+		for _, rc := range revCstr {
+			keys = append(keys, rc.validationSetKey)
```

Comment:
Just wondering, because in the `CheckPresenceInvalid()` function below you have an additional check here: don't we need to also check that `rc.Presence == asserts.PresenceRequired`?

Reply:
I don't have a strong opinion but I wonder whether we should add only the keys for validation sets that actually bind the revision. About @mardy question:  we can have a snap that is marked as optional by a validation-set but still binding its  revision (there converse is also true a validation-set can require a snap without binding the revision), of course at this point we don't expect any to mark the presence invalid but see my previous comment.

Reply:
I'm not sure about only adding the keys when v-s actually binds the revision yet, need to think.

Reply:
we should check with the store whether having validation-sets listed that don't bind the revision is a problem or not

Reply:
I asked store for clarification; FWTW I think we could  filter those out after calling this function, before calling the store.

Reply:
this seems fine then, let's remember to explore the case in the spread tests though